src/cmd/flux-job.c

```diff
@@ -919,6 +922,87 @@ void attach_signal_cb (flux_reactor_t *r, flux_watcher_t *w,
     }
 }
 
+static void attach_send_shell_completion (flux_future_t *f, void *arg)
+{
+    /* failng to write stdin to service is a fatal error */
+    if (flux_future_get (f, NULL) < 0) {
+        /* stdin may not be accepted for multiple reasons
+         * - job has completed
+         * - user requested stdin via file
+         * - stdin stream already closed due to prior pipe in
+         */
+        if (errno == ENOSYS)
+            log_msg_exit ("stdin not accepted by job");
+        log_err_exit ("attach_send_shell");
+    }
+    flux_future_destroy (f);
+}
+
+static int attach_send_shell (struct attach_ctx *ctx,
+                              void *buf,
+                              int len,
+                              bool eof)
+{
+    json_t *context = NULL;
+    char topic[1024];
+    flux_future_t *f = NULL;
+    int saved_errno;
+    int rc = -1;
+
+    snprintf (topic, sizeof (topic), "shell-%ju.stdin", (uintmax_t)ctx->id);
+    if (!(context = ioencode ("stdin", "all", buf, len, eof)))
+        goto error;
+    if (!(f = flux_rpc_pack (ctx->h, topic, ctx->leader_rank, 0, "O", context)))
+        goto error;
+    if (flux_future_then (f, -1, attach_send_shell_completion, ctx) < 0)
+        goto error;
+    /* f memory now in hands of attach_send_shell_completion() */
+    f = NULL;
+    rc = 0;
+ error:
+    saved_errno = errno;
+    json_decref (context);
+    flux_future_destroy (f);
+    errno = saved_errno;
+    return rc;
+}
+
+/* Handle std input from user */
+void attach_stdin_cb (flux_reactor_t *r, flux_watcher_t *w,
+                      int revents, void *arg)
+{
+    struct attach_ctx *ctx = arg;
+
+    if (revents & FLUX_POLLIN) {
+        ssize_t n;
+        long ps = sysconf (_SC_PAGESIZE);
+        char buf[ps];
+
+        assert (ps > 0);
+
+        if ((n = read (STDIN_FILENO, buf, ps)) < 0)
+            log_err_exit ("read on stdin");
```

Comment:
Instead of exiting on reading of stdin we may want to just log an error and stop the stdin watcher. I can't think of how this would happen (perhaps EPIPE?), but probably best not to exit with failure (?)

Reply:
I can see your point.  I guess I viewed it as "fatal" b/c this is `flux job`, not the shell.  So it's more ok to exit on an unexpected very-bad-ish error?

Reply:
Hm, good question. I leave it up to you then. I think you're right that most cmdline utilities just kind of give up on stdin read error.